enhancements/olm-crd-conversion-webhooks.md

```diff
@@ -0,0 +1,102 @@
+
+## Proposal
+
+Operator authors will be able to:
+
+- Ship their operator with a CRD conversion webhook without implementing logic to create or manage cert information as OLM will take care of that requirement instead.
+
+- Define webhook within CSV 
+
+The resources created would be similar to that created when using Admission Webhooks:
+- Webhook
+- Service
+- Deployment
+- RBAC
+- Certs (Present in the Webhook and Deployment)
+
+### CSV support for CRD Conversion Webhook
+
+The WebhookDescription struct would include one extra field called ‘ConversionCrd’ in the WebhookDescription struct which would hold the CRD name. 
+
+```
+type WebhookDescription struct {
+
+    GenerateName string `json:"generateName"`
+    Type                    WebhookAdmissionType                            `json:"type"`
+    DeploymentName          string                                          `json:"deploymentName,omitempty"`
+    ContainerPort           int32                                           `json:"containerPort,omitempty"`
+    Rules                   []admissionregistrationv1.RuleWithOperations    `json:"rules,omitempty"`
+    FailurePolicy           *admissionregistrationv1.FailurePolicyType      `json:"failurePolicy,omitempty"`
+    MatchPolicy             *admissionregistrationv1.MatchPolicyType        `json:"matchPolicy,omitempty"`
+    ObjectSelector          *metav1.LabelSelector                           `json:"objectSelector,omitempty"`
+    SideEffects             *admissionregistrationv1.SideEffectClass        `json:"sideEffects"`
+    TimeoutSeconds          *int32                                          `json:"timeoutSeconds,omitempty"`
+    AdmissionReviewVersions []string                                        `json:"admissionReviewVersions"`
+    ReinvocationPolicy      *admissionregistrationv1.ReinvocationPolicyType `json:"reinvocationPolicy,omitempty"`
+    WebhookPath             *string                                         `json:"webhookPath,omitempty"`
+    // ConversionCrd must match the CustomResourceDefinition name
+    ConversionCrd           string                                          `json:"conversionCrd"`
+}
+```
+
+Cluster admins will not be allowed to install two different versions of an Operator when one of them ships a conversion webhook. 
+There can’t be multiple conversion webhooks on a single cluster for a given CRD.
+
+### Securing webhook via certificates
+
+The certificate creation and lifecycle management is planned to utilize the same certs package that is already in OLM managing the certificates for api services. 
+
+### Deployment Installation Order
+
+OLM would:
+
+- Create Deployments defined in the CSV in order.
+- When creating the deployments defined in the CSV:
+  - Waiting for the Deployment to reach the `READY` state.
+  - Checking if the deployment owns any webhook or APIServices resources and if so, create the required resource.
```

Comment:
These resources are currently created before the deployment is ready, are we changing that?